### PR TITLE
[otbn] Refactor and speed up DIF smoke test

### PR DESCRIPTION
**Pull request by imphil**
The OTBN DIF smoke test runs two small OTBN applicatons, a barrett384 multiplication, and an error code test. These tests took around 10 minutes in CI, and around 2 minutes on reasonably new developer machines. The main purpose of this PR is to reduce the time taken for this test.

Investigations showed that the most effective way to speed up the tests is the removal of logging calls. These calls go through a simulated UART driver in Verilator simulations, which requires many cycles to be simulated. #4887 has been filed to find a system-wide solution for this problem.

In the meantime, this PR removes all "informative" logging from the DIF smoke test.

In particular, this PR contains two commits:

* First, a refactoring to use the newly merged OTBN driver for the smoke test functionality instead of calling the DIF directly, and providing all associated infrastructure within the smoke test. The main source of code duplication was the loading of applications, which is now used from the driver. This commit makes no attempt to change the functionality of the test. Functionality which isn't present in the driver has thus not been modified. One example is the writing of and reading from DMEM, where the driver provides helpers if data symbols are available, which isn't the case for the current barrett384 assembly code. This will cleaned up further in a follow-up.

* Second, the actual removal of logging.

### CODE REVIEW

**Review by rswarbrick — APPROVED** [review]
Looks good to me. And it's a good sign that switching to the driver lets you delete so much!

File: sw/device/tests/dif/dif_otbn_smoketest.c [hunk 1]
```diff
@@ -298,37 +110,24 @@ static void test_barrett384(void) {
...
+static void test_err_test(otbn_t *otbn_ctx) {
+  CHECK(otbn_load_app(otbn_ctx, kOtbnAppErrTest) == kOtbnOk);
 
-  int t_end = ibex_mcycle_read();
-  LOG_INFO("Function execution on OTBN took %d cycles (end-to-end).",
-           t_end - t_start);
-
-  check_otbn_err_code(kDifOtbnErrCodeBadDataAddr);
+  CHECK(otbn_call_function(otbn_ctx, kOtbnFuncErrTestWrapErrTest) == kOtbnOk);
+  CHECK(otbn_busy_wait_for_done(otbn_ctx) == kOtbnExecutionFailed);
```
Comment (GregAC) [inline sw/device/tests/dif/dif_otbn_smoketest.c:117]: Can we keep the check the expected error code is seen (expected error bit is set once my error bit PR is merged)?

In particular for the error bits there's a few points where things can get mixed up, we do have checking in various places for this but this is the only end to end test that can show a particular error in OTBN execution ends up setting the correct error bit as visible from Ibex.

**Review by GregAC — CHANGES_REQUESTED** [review]
I don't think we can do the check I'd like with the current driver, maybe add functionality to obtain the error code from `otbn_busy_wait_for_done`? Presumably this will be needed anyway as at some point some SW person will be wanting a more specific indication than 'fail' from the OTBN runs when they're bringing something up.

File: sw/device/tests/dif/dif_otbn_smoketest.c (hunk 1 above)
Comment (imphil) [reply sw/device/tests/dif/dif_otbn_smoketest.c:117]: I'd expect in most cases the error code to be irrelevant to software, hence the optimization for this case. I see two solutions:

* Explicitly query the error code using dif_otbn_get_err_code(). That's already possible today and can be used in the test if you'd like to. That'll cause a second read of the ERR_CODE register (right now, we could optimize this behind the scenes).
* 
* Provide an out argument to otbn_call_function() to (optionally) return the error code.

I'd go with the first option for now until we have more software written and a better understanding of what we actually need for that.

Comment (GregAC) [reply sw/device/tests/dif/dif_otbn_smoketest.c:117]: > I'd go with the first option for now

Works for me

**Comment by imphil** [comment]
This PR brings the CI runtime for the OTBN tests in Verilator down from 11 to 3 minutes, I'd call that good enough for now.

File: sw/device/tests/dif/dif_otbn_smoketest.c [hunk 2]
```diff
@@ -2,210 +2,33 @@
...
-//
-// TODO: Think about macro-ifying or auto-generating some of this code in a
-// future OTBN driver.
-extern const char _otbn_app_barrett384__imem_start[];
-extern const char _otbn_app_barrett384__imem_end[];
-extern const char _otbn_app_barrett384__dmem_start[];
-extern const char _otbn_app_barrett384__dmem_end[];
-extern const char _otbn_app_barrett384_wrap_barrett384[];
+static const otbn_app_t kOtbnAppBarrett384 = OTBN_APP_T_INIT(barrett384);
+static const otbn_ptr_t kOtbnFuncBarrett384WrapBarrett384 =
```
Comment (mcy) [inline sw/device/tests/dif/dif_otbn_smoketest.c:18]: Is there a reason for the super long constant name...?

File: sw/device/tests/dif/dif_otbn_smoketest.c [hunk 3]
```diff
@@ -2,210 +2,33 @@
...
-static const otbn_func_t kOtbnFuncErrTestWrapErrTest = {
-    .app = &kOtbnAppErrTest,
-    .func = _otbn_app_err_test_err_test,
-};
+static const otbn_app_t kOtbnAppErrTest = OTBN_APP_T_INIT(err_test);
+static const otbn_ptr_t kOtbnFuncErrTestWrapErrTest =
+    OTBN_PTR_T_INIT(err_test, err_test);
 
 const test_config_t kTestConfig = {
     .can_clobber_uart = false,
```
Comment (mcy) [inline sw/device/tests/dif/dif_otbn_smoketest.c:29]: This is the implicit default. You can just write `= {};`. Perhaps worth moving closer to the main function.

File: sw/device/tests/dif/dif_otbn_smoketest.c [hunk 4]
```diff
@@ -298,37 +110,24 @@ static void test_barrett384(void) {
...
       .base_addr = mmio_region_from_addr(TOP_EARLGREY_OTBN_BASE_ADDR),
   };
-  dif_otbn_result_t rv = dif_otbn_init(&otbn_config, &otbn);
-  CHECK(rv == kDifOtbnOk, "dif_otbn_init() failed: %d", rv);
 
-  LOG_INFO("Running barrett384 code on OTBN");
-  test_barrett384();
+  otbn_t otbn_ctx;
+
+  CHECK(otbn_init(&otbn_ctx, otbn_config) == kOtbnOk);
```
Comment (mcy) [inline sw/device/tests/dif/dif_otbn_smoketest.c:127]: Perhaps these statements want to be in one stanza together/

**Comment by imphil** [comment]
Rebased on top of the error bitfield code and addressed comments by mcy. 

@GregAC PTAL

**Review by GregAC — APPROVED** [review]
LGTM, thanks for doing this. On top of the speed-up it's good the driver is getting a smoke test too.

**Pull request merged**